### PR TITLE
Add basic attempt at Resync verisons button

### PR DESCRIPTION
**Pull request by ericholscher**
This is a bit hacky but works for me locally.
Could use some UI/UX feedback to make it work better,
I mostly just copied what was being used in other places.

Fixes https://github.com/readthedocs/ext-theme/issues/65

Screenshot 2024-09-11 at 3 00 26 PM

### CODE REVIEW

File: src/js/modules/menus.js [hunk 1]
```diff
@@ -107,3 +107,22 @@ customElements.define(
   "readthedocs-menu-build-rebuild",
   MenuBuildRebuildElement,
 );
+
+/** Menu item for resyncing all **/
+class MenuSyncAllVersionsElement extends APIEventWrapper {
+  // TODO translate this here or use a different pattern
+  static errorMessage = "There was an error syncing versions";
```
Comment (humitos) [inline src/js/modules/menus.js:114]: I always struggle with this, but I guess it's better to have final dots in sentences.

```suggestion
  static errorMessage = "There was an error syncing versions.";
```

File: readthedocsext/theme/templates/builds/partials/version_list.html [hunk 2]
```diff
@@ -17,10 +17,26 @@
...
-    <a href="{% url "project_version_create" project.slug %}" class="ui primary button">
-      <i class="fa-solid fa-plus icon"></i>
-      <span class="text">{% trans "Add version" %}</span>
-    </a>
+    <div class="right aligned eight wide computer eight wide tablet sixteen wide left aligned mobile middle aligned column">
+      {% blocktrans trimmed asvar text_resync %}
+        Resync versions
+      {% endblocktrans %}
+      <readthedocs-menu-sync-all-versions
+                                 class="ui primary button"
```
Comment (humitos) [inline readthedocsext/theme/templates/builds/partials/version_list.html:25]: I would make it more discrete here, since it's not the primary action of this page.

```suggestion
                                 class="ui button"
```

File: src/js/modules/menus.js [hunk 3]
```diff
@@ -107,3 +107,22 @@ customElements.define(
...
+
+  getSuccessURL() {
+    // Refresh the existing page to hopefully show new versions in 3 seconds, so they have time to sync.
+    // This is a bit of a hack, but it's better than nothing.
```
Comment (humitos) [inline src/js/modules/menus.js:118]: 😄

**Comment by ericholscher** [comment]
I've updated this to put it in the activate a new version page.

Screenshot 2024-09-18 at 2 31 07 PM

**Comment by agjohnson** [comment]
I would reserve that placeholder space for onboarding content. All other views use this space to point the user to more information in our docs/etc, and the button feels quite prominent for what it's function is -- manually resyncing is great to have in our UI but it is still an outlier in UX.

I'd instead reuse the existing pattern from the repository listing I pointed out above

image

https://github.com/readthedocs/ext-theme/blob/8239442a87d657aa0ac82aff9de92a73e556c838/readthedocsext/theme/templates/projects/partials/project_create_automatic.html#L101-L112

File: src/js/modules/menus.js [hunk 4]
```diff
@@ -107,3 +107,18 @@ customElements.define(
...
+  // TODO translate this here or use a different pattern
+  static errorMessage = "There was an error syncing versions.";
+
+  getSuccessURL() {
+    console.log("Successfully synced versions");
+  }
+}
+
+customElements.define(
+  "readthedocs-button-sync-all-versions",
```
Comment (agjohnson) [inline src/js/modules/menus.js:122]: You will need to update our djlint rules to support more hyphens:

https://github.com/readthedocs/ext-theme/blob/8239442a87d657aa0ac82aff9de92a73e556c838/.djlintrc#L6

Or shorten this to 3 hyphens. We tried a regex and djlint had trouble using this to match multiple levels of hyphenation.

**Comment by ericholscher** [comment]
Something like this:

Screenshot 2024-09-24 at 2 19 18 PM

**Comment by agjohnson** [comment]
That's pretty close, but looks like you added the element inside the `.ui.placeholder.segment` still. I'd check out that example from above, it shows the structure of the elements:

https://github.com/readthedocs/ext-theme/blob/8239442a87d657aa0ac82aff9de92a73e556c838/readthedocsext/theme/templates/projects/partials/project_create_automatic.html#L101-L112

In that, the `.ui.bottom.attached.message` is sibling to the outer `.ui.basic.segment > .ui.placeholder.segment`, but the `.ui.basic.segment` is only for hiding this placeholder block anyways.

**Comment by ericholscher** [comment]
This class structure is super finicky... I don't feel like I'm really learning anything about why it should actually work this way, and it's just futzing with stuff without any real direction. Is there any context for this structure?

**Comment by agjohnson** [comment]
The example I've linked to had the structure we want to replicate and copy. It uses two divs, summarized as:

```
<div class="ui top attached segment"></div>
<div class="ui bottom attached segment"></div>
```

This gives two _attached_ segments -- these have no spacing between them and they have a border separating them.

You can read about this here:

https://fomantic-ui.com/elements/segment.html#attached

Your last attempt was, summarized:

```html
<div class="ui top attached segment">
  <div class="ui placeholder segment">
    <div class="ui bottom attached segment">
    </div>
  </div>
</div>
```

This will render the bottom attached segment inside the placeholder segment, which has its own separate padding and border. The two attached segments need to be sibling elements instead.

**Comment by agjohnson** [comment]
What was here was close, but needed [e9a17d8](https://github.com/readthedocs/ext-theme/pull/481/commits/e9a17d8742394ba7a421602bc70568b48c4c6e91)

With that, it renders correctly, and matches the repository listing UI:

image

**Comment by ericholscher** [comment]
> The example I've linked to had the structure we want to replicate and copy. It uses two divs, summarized as:
> 
> ```
> <div class="ui top attached segment"></div>
> <div class="ui bottom attached segment"></div>
> ```
> 
> This gives two _attached_ segments -- these have no spacing between them and they have a border separating them.
> 
> You can read about this here:
> 
> https://fomantic-ui.com/elements/segment.html#attached
> 
> Your last attempt was, summarized:
> 
> ```
> <div class="ui top attached segment">
>   <div class="ui placeholder segment">
>     <div class="ui bottom attached segment">
>     </div>
>   </div>
> </div>
> ```
> 
> This will render the bottom attached segment inside the placeholder segment, which has its own separate padding and border. The two attached segments need to be sibling elements instead.
> 
> Placeholder segments are described here:
> 
> https://fomantic-ui.com/elements/segment.html#placeholder-segment

Nice -- we should drop this in the frontend docs somewhere, it's useful context 💯

**Comment by humitos** [comment]
I think we should add some kind of feedback or message to the user once you press the button. It shows the spinner and it stops immediately. I'd say that we should probably put a time for 2 or 3 seconds, and add a message saying "Synced, your version should appear in the list now" or similar.

Peek 2024-10-09 16-23
